File: determinant-calculator-python/calculator.py
#!/usr/bin/env python3

def prompt_for_input(prompt, validate, on_error='invalid arg'):
    tmp = input(prompt)
    if not validate(tmp):
        print(on_error)
        prompt_for_input(prompt, on_error, validate)
    return tmp


def is_int(arg):
    try:
        int(arg)
        return True
    except ValueError:
        return False


def is_float(arg):
    try:
        float(arg)
        return True
    except ValueError:
        return False


def get_minor(matrix, i, j):
    return [row[:j] + row[j + 1:] for row in (matrix[:i] + matrix[i + 1:])]


def determinant(matrix):
    n = matrix.__len__()
    assert n > 1
    if n == 2:
        return matrix[0][0] * matrix[1][1] - matrix[0][1] * matrix[1][0]

    det = 0.0
    for i in range(n):
        minor = get_minor(matrix, 0, i)
        det += (1 if i % 2 == 0 else -1) * matrix[0][i] * determinant(minor)
    return det


def main():
    rank = int(prompt_for_input('Matrix rank: ', lambda x: is_int(x)))
    matrix = [[0 for _ in range(rank)] for _ in range(rank)]
    for i, j in ((i, j) for i in range(rank) for j in range(rank)):
        matrix[i][j] = float(prompt_for_input(f"a[{i + 1}, {j + 1}] = ", lambda x: is_float(x)))

    print(f"det(A) = {determinant(matrix)}")


if __name__ == '__main__':
    main()

File: determinant-calculator/src/calculator.cpp
#include <vector>
#include <cassert>

using Matrix = std::vector<std::vector<double> >;

namespace Laplace {
    Matrix getMinor(const Matrix &matrix, const uint32_t &row, const uint32_t &col) {
        const auto n = matrix.size();
        Matrix minor(n - 1, std::vector<double>(n - 1));
        for (auto i = 0, mi = 0; i < n; ++i) {
            if (i == row) {
                continue;
            }

            for (auto j = 0, mj = 0; j < n; ++j) {
                if (j == col) {
                    continue;
                }
                minor[mi][mj] = matrix[i][j];
                mj++;
            }
            mi++;
        }
        return minor;
    }

    double determinant(const Matrix &matrix) {
        const auto n = matrix.size();
        assert(n > 1);
        if (n == 2) {
            return matrix[0][0] * matrix[1][1] - matrix[0][1] * matrix[1][0];
        }

        auto det = 0.0;
        for (auto i = 0; i < n; i++) {
            auto minor = getMinor(matrix, 0, i);
            det += (i % 2 == 0 ? 1 : -1) * matrix[0][i] * determinant(minor);
        }
        return det;
    }
}


